replay_provider.go

```diff
@@ -1,48 +1,136 @@
 package sse
 
 import (
+	"errors"
+	"strconv"
+	"sync"
 	"time"
 )
 
+// NewFiniteReplayProvider creates a finite replay provider with the given max
+// count and auto ID behaviour.
+//
+// Count is the maximum number of events FiniteReplayProvider should hold as
+// valid. It must be greater than zero.
+//
+// AutoIDs configures FiniteReplayProvider to automatically set the IDs of
+// events.
+func NewFiniteReplayProvider(
+	count int, autoIDs bool,
+) (*FiniteReplayProvider, error) {
+	if count < 1 {
+		return nil, errors.New("count must be greater than zero")
+	}
+
+	return &FiniteReplayProvider{
+		cap:     count,
+		buf:     make([]messageWithTopics, count),
+		count:   count,
+		autoIDs: autoIDs,
+	}, nil
+}
+
 // FiniteReplayProvider is a replay provider that replays at maximum a certain number of events.
 // The events must have an ID unless the AutoIDs flag is toggled.
 type FiniteReplayProvider struct {
-	b buffer
-
-	// Count is the maximum number of events FiniteReplayProvider should hold as valid.
-	// It must be a positive integer, or the code will panic.
-	Count int
-	// AutoIDs configures FiniteReplayProvider to automatically set the IDs of events.
-	AutoIDs bool
+	mut       sync.RWMutex
+	cap       int
+	buf       []messageWithTopics
+	head      int
+	tail      int
+	count     int
```

Comment:
Just wondering here: couldn't `cap(f.buf)` be used instead of a separate `cap` field? The buffer is not reallocated anywhere, neither with `make` nor with `append`. Also, `count` seems unused

Reply:
Yeah, count might be something I planned to use, then never did.

As the capacity doesn't change there's no need to look it up with cap(buf), so I stored it in a field instead. That way it can't be affected by a sneaky capacity increase bug like the original implementation either.